<a href="https://colab.research.google.com/github/iraidaantropova/final_project/blob/main/Final_Project_MLmodel_SinitsaI.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
#использую датасет "Набор данных для прогнозирования сердечной недостаточности"
#https://www.kaggle.com/datasets/andrewmvd/heart-failure-clinical-data/data

#импорт библиотек

import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split

from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier

from sklearn.metrics import confusion_matrix , classification_report , accuracy_score

Описание набора данных:

Age : возраст пациента в годах;

Anaemia : болеет ли пациент анемией;

Сreatinine_phosphokinase: уровень фермента в креатининфосфокиназа [ед/литр];

Diabetes : есть ли у пациента заболевание диабет;

Ejection_fraction : фракция выброса крови [%];

High_blood_pressure : страдает ли повышенным давлением пациент [да/нет];

Platelets : уровень тромбоцитов [тыс/мкл];

Serum_creatinine : уровень сывороточного креатинина [мкмоль/л];

Serum_sodium : уровень натрия в сыворотка [Ммоль/л ];

Sex : пол пациента;

Smoking : курение [да /нет];

Time : период наблюдения ;

DEATH_EVENT : наступила ли смерть за период наблюдения за пациентом.

In [ ]:
df=pd.read_csv('heart_failure_clinical_records_dataset.csv')
df

age  anaemia  creatinine_phosphokinase  diabetes  ejection_fraction  \
0    75.0        0                       582         0                 20   
1    55.0        0                      7861         0                 38   
2    65.0        0                       146         0                 20   
3    50.0        1                       111         0                 20   
4    65.0        1                       160         1                 20   
..    ...      ...                       ...       ...                ...   
294  62.0        0                        61         1                 38   
295  55.0        0                      1820         0                 38   
296  45.0        0                      2060         1                 60   
297  45.0        0                      2413         0                 38   
298  50.0        0                       196         0                 45   

     high_blood_pressure  platelets  serum_creatinine  serum_sodium  sex  \
0                      1  265000.00               1.9           130    1   
1                      0  263358.03               1.1           136    1   
2                      0  162000.00               1.3           129    1   
3                      0  210000.00               1.9           137    1   
4                      0  327000.00               2.7           116    0   
..                   ...        ...               ...           ...  ...   
294                    1  155000.00               1.1           143    1   
295                    0  270000.00               1.2           139    0   
296                    0  742000.00               0.8           138    0   
297                    0  140000.00               1.4           140    1   
298                    0  395000.00               1.6           136    1   

     smoking  time  DEATH_EVENT  
0          0     4            1  
1          0     6            1  
2          1     7            1  
3          0     7            1  
4          0     8            1  
..       ...   ...          ...  
294        1   270            0  
295        0   271            0  
296        0   278            0  
297        1   280            0  
298        1   285            0  

[299 rows x 13 columns]

In [ ]:
#Цель исследования — создать модель, которая может успешно предсказать наступление смерти( положительный случай (1) в датасете) с вероятностью выше,
#чем случайное угадывание (0,5).
#для создания модели делю данные на обучающие и тестовые

x = df.iloc[:,0:11].values
y = df['DEATH_EVENT'].values

x_train, x_test, y_train, y_test= train_test_split(x,y,test_size=0.2,shuffle=True,random_state=42)

In [ ]:
#стандартизирую числовые переменные, чтобы все переменные вносили одинаковый вклад в создание модели

st_x= StandardScaler()
x_train= st_x.fit_transform(x_train)
x_test= st_x.transform(x_test)

In [ ]:
#Логистическая регрессия
lg=LogisticRegression(C=5)
lg.fit(x_train,y_train)

LogisticRegression(C=5)

In [ ]:
lg.score(x_train,y_train)

0.7782426778242678

In [ ]:
y_pred=lg.predict(x_test)


In [ ]:
#точность модели(общее количество правильных прогнозов модели)
acc = accuracy_score(y_test, y_pred)
print("Точность модели логистической регрессии составляет (в %):", acc*100)


Точность модели логистической регрессии составляет (в %): 65.0


In [ ]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.64      0.91      0.75        35
           1       0.70      0.28      0.40        25

    accuracy                           0.65        60
   macro avg       0.67      0.60      0.58        60
weighted avg       0.67      0.65      0.61        60



Целью модели было предсказать, умрет ли кто-то из-за сердечной недостаточности или нет, в данном случае важно успешно прогнозировать положительные случаи (1) по сравнению с фактическими положительными случаями.

Судя по отчету о классификации можно сделать вывод, что из всех случаев со смертельным исходом по прогнозу модели, только в 70% на самом деле так и было
Из всех случаев со смертельным исходом, которые действительно были, модель правильно предсказала так на 28% случаев

Значение F1 -  средневзвешенное гармоническое значение точности и полноты, чем ближе к 1, тем лучше модель.
В данном случае значение 0.40, что не очень близко к 1, это говорит о том, что модель плохо справляется с прогнозом


In [ ]:
#случайный лес (RandomForestClassifier)

Rf=RandomForestClassifier()
Rf.fit(x_train,y_train)

RandomForestClassifier()

In [ ]:
Rf.score(x_train,y_train)

1.0

In [ ]:
y_pred=Rf.predict(x_test)

In [ ]:
#точность модели(общее количество правильных прогнозов модели)
acc = accuracy_score(y_test, y_pred)
print("Точность модели RandomForestClassifier составляет (в %):", acc*100)

Точность модели RandomForestClassifier составляет (в %): 70.0


In [ ]:
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.70      0.86      0.77        35
           1       0.71      0.48      0.57        25

    accuracy                           0.70        60
   macro avg       0.70      0.67      0.67        60
weighted avg       0.70      0.70      0.69        60



Судя по отчету о классификации в этой модели делаю вывод, что из всех случаев со смертельным исходом по прогнозу модели, в 71% на самом деле так и было, что на 1% выше, чем у предыдущей модели
Из всех случаев со смертельным исходом, которые действительно были, модель правильно предсказала так на 48% случаев, что также больше, чем у модели логистической регрессии на 20%.

Значение F1 -  средневзвешенное гармоническое значение точности и полноты, чем ближе к 1, тем лучше модель.
В данном случае значение 0.57, что гораздо ближе к 1, это говорит о том, что модель лучше справляется с прогнозом.

Рассмотрев 2 модели, можно увидеть, что оценка отзыва положительного случая (1) в 1 модели  составляет 0,28, что очень мало, во 2 модели - 0.48, что, безусловно,   гораздо более высокая оценка отзыва, но меньше, чем  вероятность случайного угадывания, это говорит о том, что данные модели использовать нецелесообразно